src/platform/Linux/bluez/MainLoop.cpp

```diff
@@ -0,0 +1,189 @@
+
+#if CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE
+
+#include <errno.h>
+#include <pthread.h>
+#include <support/CodeUtils.h>
+#include <support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace DeviceLayer {
+namespace Internal {
+
+namespace {
+
+class Semaphore
+{
+public:
+    Semaphore()
+    {
+        if (sem_init(&mSemaphore, 0 /* shared */, 0 /*value*/) != 0)
+        {
+            ChipLogError(DeviceLayer, "Failed to initialize semaphore.");
+        }
+    }
+
+    ~Semaphore()
+    {
+        if (sem_destroy(&mSemaphore) != 0)
+        {
+            ChipLogError(DeviceLayer, "Failed to destroy semaphore.");
+        }
+    }
+
+    void Post() { sem_post(&mSemaphore); }
+
+    void Wait() { sem_wait(&mSemaphore); }
+
+private:
+    sem_t mSemaphore;
+};
+
+int PostSemaphore(Semaphore * semaphore)
+{
+    semaphore->Post();
+    return 0;
+}
+
+class CallbackIndirection
+{
+public:
+    CallbackIndirection(GSourceFunc f, void * a) : mCallback(f), mArgument(a) {}
+
+    void Wait() { mDoneSemaphore.Wait(); }
+
+    static int Callback(CallbackIndirection * self)
+    {
+        int result = self->mCallback(self->mArgument);
+        self->mDoneSemaphore.Post();
+        return result;
+    }
+
+private:
+    Semaphore mDoneSemaphore;
+    GSourceFunc mCallback;
+    void * mArgument;
+};
+
+} // namespace
+
+MainLoop & MainLoop::Instance()
+{
+    static MainLoop sMainLoop;
+    return sMainLoop;
+}
+
+void * MainLoop::Thread(void * self)
+{
+    MainLoop * loop = reinterpret_cast<MainLoop *>(self);
+
+    ChipLogDetail(DeviceLayer, "TRACE: Bluez mainloop starting %s", __func__);
+    g_main_loop_run(loop->mBluezMainLoop);
+    ChipLogDetail(DeviceLayer, "TRACE: Bluez mainloop stopping %s", __func__);
+
+    if (loop->mCleanup != nullptr)
+    {
+        ChipLogDetail(DeviceLayer, "TRACE: Executing cleanup %s", __func__);
+        loop->mCleanup(loop->mCleanupArgument);
+    }
+
+    return nullptr;
+}
+
+CHIP_ERROR MainLoop::EnsureStarted()
+{
+    if (mBluezMainLoop != nullptr)
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    mBluezMainLoop = g_main_loop_new(nullptr, TRUE);
+    if (mBluezMainLoop == nullptr)
+    {
+        ChipLogError(DeviceLayer, "FAIL: memory alloc in %s", __func__);
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    int pthreadErr = pthread_create(&mThread, nullptr, &MainLoop::Thread, reinterpret_cast<void *>(this));
+    int tmpErrno   = errno;
+    if (pthreadErr != 0)
+    {
+        ChipLogError(DeviceLayer, "FAIL: pthread_create (%s) in %s", strerror(tmpErrno), __func__);
+        g_free(mBluezMainLoop);
+        mBluezMainLoop = nullptr;
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    Semaphore semaphore;
+
+    GMainContext * context = g_main_loop_get_context(mBluezMainLoop);
+    if (context == nullptr)
+    {
+        ChipLogError(DeviceLayer, "Unexpected null context on the main loop");
+        // bluez loop not cleaned as running thread is already started.
+        return CHIP_ERROR_INTERNAL;
```

Comment:
free the main loop?

Reply:
I cannot free it because the underlying thread that will do 'run main loop' on it will be active.

We could add more logic to fix it up (maybe by schedule-and-wait on something that kills the main thread) however this seems impossible to really happen at runtime: we just created the object and it has to have a context. 

If we need to update it, I would prefer to do it separately or 'VerifyOrDie' instead to not worry about the case.

Reply:
Updated to VerifyOrDie